### ISSUE TITLE
Adopt GitVersion

### ISSUE THREAD

**Issue by adamralph** [FakeItEasy/FakeItEasy#1046]
This will remove the need for:

- `VERSION_SUFFIX` to build pre-releases
- bumping the minor version or patch (only major bumps required)

Depends on #955.

**Comment by thomaslevesque** [comment]
Why not, but I have no idea how it works. Do you have any experience with GitVersion @adamralph?

**Comment by adamralph** [comment]
@thomaslevesque yes, we use it for all our NServiceBus packages. It's [playfully simple to use]( http://gitversion.readthedocs.io/en/latest/usage/msbuild-task/).

**Comment by blairconrad** [comment]
And super simple to add [even if you don't use the MSBuild integration](https://github.com/blairconrad/SelfInitializingFakes/pull/16)!

**Comment by blairconrad** [comment]
Taking off-hold, since #955 is fixed.

**Comment by blairconrad** [comment]
So I naively incorporated the task and built.

> Running '"./.nuget/NuGet.exe" pack ./src\FakeItEasy.Analyzer.nuspec -Version 3.6.0-gitversion.1+1558.Branch.gitversion.Sha.07999377693257e18799a14b9e90e43fa6390f42 -OutputDirectory ./artifacts/output -NoPackageAnalysis'

I'm not so sure about the 3.6.0. Will investigate.

Aha! From build.log:

>     INFO [05/05/17 7:38:37:94] Merge message 'Merge remote-tracking branch 'remotes/origin/3.5'': 3.5.0 with commit count source 0b7d492dd1202b9172596ac3eafb87356c92c8f8 (Incremented: 3.6.0)

It found an old branch named 3.5 and decided that it was a version. Terrifying.

**Comment by blairconrad** [comment]
The 3.5 still wins even if we're on a `release/3.4.0` branch (or have merged one in), but a tag of 3.4.0 would cause the build to come out with release 3.4.0.

**Comment by blairconrad** [comment]
@adamralph  (not to ignore you, @thomaslevesque, but you and are I definitely less experienced here), does any of this surprise you?
Folks, I guess having the "develop" get labelled with oddly-high 3.6.0 versions for a release or two (let's release quickly) isn't so bad, as long as we get good versions out once we tag, but it seems that we run a small risk of recurrence if someone uses a numeric branch in the future.

Also, if we switch to GitVersion, I think we would then have no tasks to perform on the release branches (for most releases, at least).
Then the merge into master is effectively merging develop in, and the merge of the feature branch into develop is a no-op. In order to get the version bump on develop, and give an indication that we included whatever was in the latest release in develop, we'd have to… merge master back into develop?

![image](https://cloud.githubusercontent.com/assets/3275797/25768257/69065cb8-31cf-11e7-87e3-834c458968a9.png)

I fear I'm missing something.

**Comment by adamralph** [comment]
What is this `3.5` branch of which you speak? I don't see any branch (or tag) name that.

To release when using GitVersion:

- create a `release-3.4.0` branch from `develop`
- push it and build it
- if you want to RC, release an RC from the `release-3.4.0` branch
- if any bugs come out of RC, fix them on the `release-3.4.0` branch
- when ready to release, send a PR from `release-3.4.0` to `master`
- when that PR is merged, the resulting package is the RTM package. Push it to NuGet.
- if any changes were made on the `release-3.4.0` branch, send a PR from `release-3.4.0` to `develop`

**Comment by blairconrad** [comment]
Sounds generally good, @adamralph, but if there's no change on the release
branch, won't develop end up producing 3.3.x builds?

The 3.5 branch may be specific to my repo. I am away from my usual
computer, so can't get the refspec, if that is the word.

**Comment by adamralph** [comment]
> won't develop end up producing 3.3.x builds?

@blairconrad no, GitVersion is clever enough to look back at all the history and work things out.

**Comment by blairconrad** [comment]
Even though no 3.4 branch is in develop's past?
We should release soon so I can see this.

**Comment by blairconrad** [comment]
@adamralph, I found the mysterious 3.5 branch by using `gitversion /diag`:

```
INFO [05/23/17 11:31:12:36] Merge message 'Merge remote-tracking branch 'remotes/origin/3.5'': 3.5.0 with commit count source 0b7d492dd1202b9172596ac3eafb87356c92c8f8 (Incremented: 3.6.0)
```

Patrik merged the 3.5 branch in commit 0b7d492dd1202b9172596ac3eafb87356c92c8f8 on 2011-10-16, so now GitVersion really likes 3.6 for our next version.
I suspect this is because "GitVersion is clever enough to look back at all the history and work things out." :wink:

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.4.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.4.0).
